varname/ignore.py

```diff
@@ -0,0 +1,360 @@
+            self.dirname = f'{self.dirname}{path.sep}'
+
+    def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
+        frame = frameinfos[frame_no].frame
+
+        return path.realpath(frame.f_code.co_filename).startswith(
+            path.realpath(self.dirname)
+        )
+
+    def __repr__(self):
+        return f'<IgnoreDirname({self.dirname!r})>'
+
+class IgnoreFunction(IgnoreElem, attrs=['func']):
+    """Ignore a non-decorated function"""
+    def _post_init(self) -> None:
+        if (
+                # without functools.wraps
+                '<locals>' in self.func.__qualname__ or
+                self.func.__name__ != self.func.__code__.co_name
+        ):
+            warnings.warn(
+                f'You asked varname to ignore function {self.func.__name__!r}, '
+                'which may be decorated. If it is not intended, you may need '
+                'to ignore all intermediate frames with a tuple of '
+                'the function and the number of its decorators.',
+                MaybeDecoratedFunctionWarning
+            )
+
+    def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
+        frame = frameinfos[frame_no].frame
+        return frame.f_code == self.func.__code__
+
+    def __repr__(self):
+        return f'<IgnoreFunction({self.func.__name__})>'
+
+class IgnoreDecorated(IgnoreElem, attrs=['func', 'n_decor']):
+    """Ignore a decorated function"""
+    def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
+        try:
+            frame = frameinfos[frame_no + self.n_decor].frame
+        except IndexError:
+            return False
+
+        return frame.f_code == self.func.__code__
+
+    def __repr__(self):
+        return f'<IgnoreDecorated({self.func.__name__}, {self.n_decor})>'
+
+class IgnoreModuleQualname(IgnoreElem, attrs=['module', 'qualname']):
+    """Ignore calls by qualified name in the module"""
+
+    def _post_init(self) -> None:
+
+        attach_ignore_id_to_module(self.module)
+        # check uniqueness of qualname
+        modfile = getattr(self.module, '__file__', None)
+        if modfile is not None:
+            check_qualname_by_source(
+                Source.for_filename(modfile, self.module.__dict__),
+                self.module.__name__,
+                self.qualname
+            )
+
+    def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
+        frame = frameinfos[frame_no].frame
+        module = cached_getmodule(frame.f_code)
+
+        # Return earlier to avoid qualname uniqueness check
+        if module and module != self.module:
+            return False
+
+        if (
+                not module and
+                not frame_matches_module_by_ignore_id(frame, self.module)
+        ):
+            return False
+
+        source = Source.for_frame(frame)
+        check_qualname_by_source(source, self.module.__name__, self.qualname)
```

Comment:
@alexmojaki 
Note the first and last lines of this block. We are checking this twice (or even more if `match` is called multiple times)

Do we really not need a cached version of `check_qualname_by_source` here (you mentioned cached-free)?
Is `Source` object good as an argument in a `lru_cache` function?

Reply:
Sorry, I was saying a free function that is cached. By free function I mean not a method, as you've done. You should use lru_cache and it will work well with Source.

Reply:
Done. Let me know if you have any other suggestions for the entire PR.